experimental/list_formatter/src/provider.rs

```diff
@@ -0,0 +1,237 @@
+    pub end: ConditionalListJoinerPattern<'data>,
+    #[cfg_attr(feature = "provider_serde", serde(borrow))]
+    pub pair: ConditionalListJoinerPattern<'data>,
+}
+
+/// A pattern that can behave conditionally on the next element.
+#[derive(Debug, PartialEq, Clone, Yokeable, ZeroCopyFrom)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub struct ConditionalListJoinerPattern<'data> {
+    #[cfg_attr(feature = "provider_serde", serde(borrow))]
+    default: ListJoinerPattern<'data>,
+    #[cfg_attr(feature = "provider_serde", serde(borrow))]
+    special_case: Option<SpecialCasePattern<'data>>,
+}
+
+/// A pattern that can behave conditionally on the next element.
+#[derive(Debug, PartialEq, Clone, Yokeable, ZeroCopyFrom)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+struct SpecialCasePattern<'data> {
+    #[cfg_attr(feature = "provider_serde", serde(borrow))]
+    condition: Cow<'data, str>, // TODO: Serialize a compiled regex instead
+    #[cfg_attr(feature = "provider_serde", serde(borrow))]
+    pattern: ListJoinerPattern<'data>,
+}
+
+/// A pattern containing two numeric placeholders ("{0}, and {1}.")
+#[derive(Debug, PartialEq, Clone, Yokeable, ZeroCopyFrom)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+struct ListJoinerPattern<'data> {
+    /// The pattern string without the placeholders
+    string: Cow<'data, str>,
+    /// The indices of the two placeholders
+    insertion_points: [usize; 2],
+}
+
+pub type PatternParts<'a> = (&'a str, &'a str, &'a str);
+
+impl<'data> ListJoinerPattern<'data> {
+    fn borrow_tuple(&'data self) -> PatternParts<'data> {
+        (
+            &self.string[0..self.insertion_points[0]],
+            &self.string[self.insertion_points[0]..self.insertion_points[1]],
+            &self.string[self.insertion_points[1]..],
+        )
+    }
+}
+
+impl<'a> ConditionalListJoinerPattern<'a> {
+    pub fn parts(&'a self, following_value: &str) -> PatternParts<'a> {
+        match &self.special_case {
+            Some(SpecialCasePattern { condition, pattern })
+                if regex::Regex::new(&*condition)
+                    .unwrap()
+                    .is_match(following_value) =>
+            {
+                pattern.borrow_tuple()
+            }
+            _ => self.default.borrow_tuple(),
+        }
+    }
+}
+
+#[cfg(any(test, feature = "provider_transform_internals"))]
+pub mod pattern_construction {
+    use super::*;
+    use crate::error::Error;
+    use core::str::FromStr;
+
+    impl<'data> FromStr for ListJoinerPattern<'data> {
+        type Err = Error;
+        fn from_str(pattern: &str) -> Result<Self, Self::Err> {
+            match (pattern.find("{0}"), pattern.find("{1}")) {
+                (Some(index_0), Some(index_1)) if index_0 + 3 <= index_1 => Ok(ListJoinerPattern {
```

Comment:
Praise: This is cool parsing code :)

Reply:
Performance wise it is slightly suboptimal as it scans the patterns twice to find those indexes.
Performance of the parser is not cirtical because we use pre-parsed patterns at runtime, but we may want to comment here and/or file an issue (help wanted, good first etc.) to improve this to scan pattern once for all placeholders.

Reply:
These strings are barely longer than 10 characters, so this really is a micro optimization

Reply:
Correct, that's the edge performance we do optimize for in many of our runtime code (one fewer loop, one fewer allocation etc.).
Since this is not on runtime path, I agree it is not important and suggested a follow up issue as good-first-bug.

Reply:
I disagree, because:

1. I'm not convinced that this is a competitive optimization relative to other low-hanging fruit we have in the transformer code
2. The transformer code should be optimized for readability/maintainability rather than micro-speed, as previously agreed in the style guide
3. Hopefully we will be able to consider changing this to `icu_pattern` anyway in the near future

Reply:
I'm convinced by 3.